packages/playwright-file-snapshots/src/matchers/file-matcher.ts

```diff
@@ -25,39 +36,86 @@ interface MatchValidationFileParams {
   matcherState: ExpectMatcherState;
 }
 
-export function matchValidationFile(
+export async function matchValidationFile(
   params: MatchValidationFileParams,
-): MatcherReturnType {
+): Promise<MatcherReturnType> {
   const { matcherName, actual, serializer, config, options, matcherState } =
     params;
-  const { isNot, utils } = matcherState;
+  const { timeout, isNot, utils } = matcherState;
 
   if (isNot) {
     throw new Error("Matcher negation is not supported");
   }
 
   const { validationDir, outputDir, filterSteps } = config;
-  const { titlePath, testPath } = parseTestInfo(test.info(), filterSteps);
-  let pass: boolean;
+  const { titlePath, testPath, updateSnapshots } = parseTestInfo(
+    test.info(),
+    filterSteps,
+  );
 
   const { name, namingStrategy } = options;
-  const matcherResult = new ValidationFileMatcher({
+  const matcher = new ValidationFileMatcher({
     validationDir,
     outputDir,
     testPath,
     titlePath,
     name,
     namingStrategy,
     serializer,
-  }).matchFileSnapshot(actual);
-  matcherResult.writeFileSnapshots();
-
-  try {
-    baseExpect(matcherResult.actual).toBe(matcherResult.expected);
-    pass = true;
-  } catch {
-    pass = false;
+  });
+
+  const testTimeout = options.timeout ?? timeout;
+  const updateDelay = Math.min(testTimeout, UPDATE_DELAY);
+  const snapshot = createSnapshotInstance(actual);
+
+  const isUpdate = matcher.isValidationFileMissing || updateSnapshots;
+  if (isUpdate && snapshot instanceof RepeatableSnapshot) {
+    await waitForTimer(updateDelay);
+  }
+
+  while (true) {
+    const currentActual = await snapshot.getValue();
+    const matcherResult = matcher.matchFileSnapshot(currentActual);
+
+    let pass: boolean;
+
+    try {
+      baseExpect(matcherResult.actual).toBe(matcherResult.expected);
+      pass = true;
+    } catch {
+      pass = false;
+    }
+
+    if (
+      pass ||
+      isUpdate ||
+      snapshot instanceof StaticSnapshot ||
+      snapshot.isTimeoutExceeded(testTimeout)
```

Comment:
Maybe extract this check into a method

Reply:
I see, I think you have a point there. I extracted the condition to a constant named `skipRetry`. I hope this makes the intention clearer.